dataset/dataset.py

```diff
@@ -153,56 +172,72 @@ def generate_aux_table_sql(self, aux_table, query, index_attrs=False):
 
     def get_raw_data(self):
         """
-        Is this guaranteed sorted by TID?
+        get_raw_data returns a pandas.DataFrame containing the raw data as it was initially loaded.
         """
-        if self.raw_data:
-            return self.raw_data.df
-        else:
+        if self.raw_data is None:
             raise Exception('ERROR No dataset loaded')
+        return self.raw_data.df
 
     def get_attributes(self):
-        if self.raw_data:
-            attrs = self.raw_data.get_attributes()
-            attrs.remove('_tid_')
-            return attrs
-        else:
+        """
+        get_attributes return the trainable/learnable attributes (i.e. exclude meta
+        columns like _tid_).
+        """
+        if self.raw_data is None:
             raise Exception('ERROR No dataset loaded')
+        return self.raw_data.get_attributes()
 
     def get_cell_id(self, tuple_id, attr_name):
         """
         get_cell_id returns cell ID: a unique ID for every cell.
 
         Cell ID: _tid_ * (# of attributes) + attr_idx
         """
-        vid = tuple_id*self.attr_number + self.attr_to_idx[attr_name]
-
+        vid = tuple_id*self.attr_count + self.attr_to_idx[attr_name]
         return vid
 
     def get_statistics(self):
+        """
+        get_statistics returns:
+            1. self.total_tuples (total # of tuples)
+            2. self.single_attr_stats ({ attribute -> { value -> count } })
+              the frequency (# of entities) of a given attribute-value
+            3. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
+              the statistics for each pair of attributes, attr1 and attr2, where:
+                <attr1>: first attribute
+                <attr2>: second attribute
+                <val1>: all values of <attr1>
+                <val2>: values of <attr2> that appear at least once with <val1>.
+                <count>: frequency (# of entities) where attr1=val1 AND attr2=val2
+        """
         if not self.stats_ready:
+            logging.debug('computing frequency and co-occurrence statistics from raw data...')
+            tic = time.clock()
             self.collect_stats()
+            logging.debug('DONE computing statistics in %.2fs', time.clock() - tic)
+
         stats = (self.total_tuples, self.single_attr_stats, self.pair_attr_stats)
         self.stats_ready = True
         return stats
 
     def collect_stats(self):
         """
         collect_stats memoizes:
-          1. self.single_attr_stats ({ attribute -> Series (value -> count) })
+          1. self.single_attr_stats ({ attribute -> { value -> count } })
             the frequency (# of entities) of a given attribute-value
-          2. self.pair_attr_stats ({ attr1 -> { attr2 -> DataFrame } } where
-            DataFrame contains 3 columns:
+          2. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
```

Comment:
I think descriptions for a few keys in the dict are missing. 
Copied from `get_statistics` docstring
```suggestion
               3. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
              the statistics for each pair of attributes, attr1 and attr2, where:
                <attr1>: first attribute
                <attr2>: second attribute
                <val1>: all values of <attr1>
                <val2>: values of <attr2> that appear at least once with <val1>.
                <count>: frequency (# of entities) where attr1=val1 AND attr2=val2
        """
```